File: Board.cpp
#include <iostream>
#include <vector>

class Board {
private:
    std::vector<std::vector<Symbol*>> grid;
    int size;
    Symbol* emptyCell;

public:
    Board(int s) : size(s), emptyCell(new Symbol('-')) {
        grid = std::vector<std::vector<Symbol*>>(size, std::vector<Symbol*>(size, emptyCell));
    }

    bool isCellEmpty(int row, int col) {
        if(row < 0 || row >= size || col < 0 || col >= size) {
            return false;
        }
        return grid[row][col] == emptyCell;
    }

    bool placeMark(int row, int col, Symbol* mark) {
        if(row < 0 || row >= size || col < 0 || col >= size) {
            return false;
        }
        if(!isCellEmpty(row, col)) {
            return false;
        }
        grid[row][col] = mark;
        return true;
    }

    Symbol* getCell(int row, int col) {
        if(row < 0 || row >= size || col < 0 || col >= size) {
            return emptyCell;
        }
        return grid[row][col];
    }

    int getSize() {
        return size;
    }

    Symbol* getEmptyCell() {
        return emptyCell;
    }

    void display() {
        std::cout << "\n  ";
        for(int i = 0; i < size; i++) {
            std::cout << i << " ";
        }
        std::cout << std::endl;

        for(int i = 0; i < size; i++) {
            std::cout << i << " ";
            for(int j = 0; j < size; j++) {
                std::cout << grid[i][j]->getMark() << " ";
            }
            std::cout << std::endl;
        }
        std::cout << std::endl;
    }
};

File: frontend/src/logic/Board.js
import { Symbol } from './Symbol.js';

export class Board {
    constructor(size) {
        this.size = size;
        this.emptyCell = new Symbol('-');
        this.grid = Array.from({ length: size }, () => Array(size).fill(this.emptyCell));
    }

    isCellEmpty(row, col) {
        if (row < 0 || row >= this.size || col < 0 || col >= this.size) {
            return false;
        }
        return this.grid[row][col] === this.emptyCell;
    }

    placeMark(row, col, mark) {
        if (row < 0 || row >= this.size || col < 0 || col >= this.size) {
            return false;
        }
        if (!this.isCellEmpty(row, col)) {
            return false;
        }
        this.grid[row][col] = mark;
        return true;
    }

    getCell(row, col) {
        if (row < 0 || row >= this.size || col < 0 || col >= this.size) {
            return this.emptyCell;
        }
        return this.grid[row][col];
    }

    getSize() {
        return this.size;
    }

    getEmptyCell() {
        return this.emptyCell;
    }
    
    // helper for UI
    getFlatBoard() {
        let flat = [];
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                flat.push(this.grid[i][j] === this.emptyCell ? null : this.grid[i][j].getMark());
            }
        }
        return flat;
    }
}
